### ISSUE TITLE
Document JobDSL Syntax for googleCloudStorageUploader Usage in publishers Context

### ISSUE THREAD

**Issue by rmeffan** [jenkinsci/google-storage-plugin#68]
Summary
=======

I have been unable to determine how to properly configure a Jenkins JobDSL equivalent to the UI-configured post-build action implementation of a Google Cloud Storage plugin "Build Log Upload" operation. Since it appears that the plugin has been implemented in a way that is compatible with the JobDSL plugin, this issue is a request for documentation or sample code for how to implement googleCloudStorageUploader post-build publisher actions using JobDSL.

Background 
---------
1. The Google Cloud Storage plugin is installed and working properly in our Jenkins instance.

2. The Jenkins JobDSL API viewer for our instance shows options for the following entities:
- googleCloudStorageUploader
- googleStorageDownload
- googleStorageUpload

Google Cloud Storage Uploader: Post-build Action (publishers context)
---------------------------------------------------------------------
1. googleCloudStorageUploader works properly as a Post-build action in the publishers context when configured via the Jenkins UI. I can add a post-build action of type "Google Cloud Storage Plugin", and add 1..N operations of type "Bucket with expiring elements lifecycle", "Build Log Upload", or "Classic Log Upload". 

2. When I expand the googleCloudStorageUploader entity in the API viewer for our instance, this is what is rendered in the documentation/sample field:
```
# Failed attempt 1 at JobDSL Equivalent
googleCloudStorageUploader {
    credentialsId(String value)
    uploads {
    }
}
```

3. If I drop this code into a test JobDSL job, and adjust the credentialsId paramter to the same value I configured in the UI, it throws the following error re: the empty uploads section:
```
# Error during JobDSL rendering of attempt 1 at JobDSL Equivalent
ERROR: Could not instantiate {credentialsId=name-of-credential, uploads=[]} for GoogleCloudStorageUploader(credentialsId: String, uploads: AbstractUpload{ClassicUpload(bucket: String, module: org.kohsuke.stapler.NoStaplerConstructorException: There's no @DataBoundConstructor on any constructor of class com.google.jenkins.plugins.storage.UploadModule, pattern: String, bucketNameWithVars: String, sourceGlobWithVars: String, forFailedJobs?: boolean, pathPrefix?: String, sharedPublicly?: boolean, showInline?: boolean) | ExpiringBucketLifecycleManager(bucket: String, module: org.kohsuke.stapler.NoStaplerConstructorException: There's no @DataBoundConstructor on any constructor of class com.google.jenkins.plugins.storage.UploadModule, ttl: int, bucketNameWithVars: String, bucketObjectTTL: int, forFailedJobs?: boolean, pathPrefix?: String, sharedPublicly?: boolean, showInline?: boolean) | StdoutUpload(bucket: String, module: org.kohsuke.stapler.NoStaplerConstructorException: There's no @DataBoundConstructor on any constructor of class com.google.jenkins.plugins.storage.UploadModule, logName: String, bucketNameWithVars: String, forFailedJobs?: boolean, pathPrefix?: String, sharedPublicly?: boolean, showInline?: boolean)}[]): java.lang.reflect.InvocationTargetException
```

4. Using a combination of the config.xml diff from before/after creating a Google Cloud Storage Uploader "Build Log Uploader" operation, and the StdoutUpload() errorr message from item 3, I constructed variants of the following JobDSL code:

```
# config.xml snippet from UI-configured Google Cloud Storage Uploader "Build Log Uploader" upload operation
<com.google.jenkins.plugins.storage.GoogleCloudStorageUploader plugin="google-storage-plugin@1.1">
  <credentialsId>name-of-credential</credentialsId>
    <uploads>
      <com.google.jenkins.plugins.storage.StdoutUpload>
      <bucketNameWithVars>gs://test-jenkins-logs/${JOB_NAME}/${BUILD_NUMBER}</bucketNameWithVars>
      <sharedPublicly>false</sharedPublicly>
      <forFailedJobs>true</forFailedJobs>
      <showInline>false</showInline>
      <module/>
      <logName>build-log.txt</logName>
    </com.google.jenkins.plugins.storage.StdoutUpload>
  </uploads>
</com.google.jenkins.plugins.storage.GoogleCloudStorageUploader>
```

```
# Failed attempt 2 at JobDSL Equivalent
    publishers {
        googleCloudStorageUploader {
            credentialsId('name-of-credential')
            uploads {
                logName('build-log.txt')
                StdoutUpload {
                    bucketNameWithVars('gs://test-jenkins-logs/${JOB_NAME}/${BUILD_NUMBER}')
                    sharedPublicly(false)
                    forFailedJobs(true)
                    showInline(false)
                }
            }
        }
    }
```

```
# Error during JobDSL rendering of attempt 2 at JobDSL Equivalent
ERROR: (FiabStartRobMTest.groovy, line 118) No signature of method: javaposse.jobdsl.plugin.structs.DescribableListContext.logName() is applicable for argument types: (java.lang.String) values: [build-log.txt]
```

```
# Failed attempt 3 at JobDSL Equivalent
    publishers {
         googleCloudStorageUploader {
            credentialsId('name-of-credential')
            uploads{
                logName('build-log.txt')
                StdoutUpload {
                    logName('build-log.txt')
                    bucketNameWithVars('gs://test-jenkins-logs/${JOB_NAME}/${BUILD_NUMBER}')
                    sharedPublicly(false)
                    forFailedJobs(true)
                    showInline(false)
                }
            }
        }
    }
```

```
# Error during JobDSL rendering of attempt 3 at JobDSL Equivalent
ERROR: (FiabStartRobMTest.groovy, line 118) No signature of method: javaposse.jobdsl.plugin.structs.DescribableListContext.StdoutUpload() is applicable for argument types: (FiabStartRobMTest$_run_closure1$_closure7$_closure23$_closure26$_closure27) values: [FiabStartRobMTest$_run_closure1$_closure7$_closure23$_closure26$_closure27@60761e6a]
```

Google Storage Upload: Freestyle Job (steps -> step context)
------------------------------------------------------------
1. I can configure a build step in a Jenkins Freestyle job

2. When I expand the googleStorageUpload entity in the API viewer for our instance, this is what is rendered in the documentation/sample field:
```
googleStorageUpload {
    credentialsId('name-of-credential')
    bucket('gs://test-jenkins-logs/${JOB_NAME}/${BUILD_NUMBER}')
    pattern('**/log')
    sharedPublicly(false)
    showInline(false)
}
```

3. If I drop this code into a test JobDSL job, and adjust the parameters to the same values I configured using the UI, the job renders properly and the result looks the same when I view the Jenkins job's configuration using the UI.

**Comment by rachely3n** [comment]
You should definitely follow the syntax of:
```
googleStorageUpload {
    credentialsId('name-of-credential')
    bucket('gs://test-jenkins-logs/${JOB_NAME}/${BUILD_NUMBER}')
    pattern('**/log')
    sharedPublicly(false)
    showInline(false)
}
```

The DSL will follow the names contained in this file: https://github.com/jenkinsci/google-storage-plugin/blob/develop/src/main/java/com/google/jenkins/plugins/storage/ClassicUploadStep.java

I'm currently writing integration tests to test pipeline DSL, so I think it'll be more obvious how to run such jobs after that's complete.

Also including another example to hopefully unblock you for the moment: https://github.com/craigatgoogle/cloudbees-days-seattle-demo/blob/eb0464564d45c91dbf3e2e560af1440d95b55938/Jenkinsfile#L36

**Comment by rmeffan** [comment]
Hi @rachely3n,

Thanks for the comment and the links.

Re: "You should definitely follow the syntax of: googleStorageUpload", does this mean that `googleCloudStorageUploader` is not a viable solution in the present if we require a JobDSL implementation of it? The reason why I'd like to use this option is because it can be executed as a post-build action. When I tried to use `googleStorageUpload`, the only JobDSL implementation that worked was to execute it as a build step. However, when I used it in that location, IIRC it couldn't see post-build actions (which makes sense, of course).

Re: the link to ClassicStorageUpload.java, I've already looked at that file, as well as the other files in https://github.com/jenkinsci/google-storage-plugin/tree/develop/src/main/java/com/google/jenkins/plugins/storage , but I'll give it another look to be sure I'm using the right names for things.

Re: the link to the example, that's pipeline syntax, but it may be helpful. I see that there's a class parameter in the invocation; that's something I thought I might need to use, given the error output during the DSL rendering, but I wasn't sure which class or DSL formatting to use. Will try a few things, and see what works.

Thanks again for taking the time to respond. Appreciate it!

**Comment by rachely3n** [comment]
My mistake, didn't realize it was Post Build Action.

If you look at the Google cloud storage uploader post build step in the Jenkins UI, you'll notice there are 3 other options, 

Bucket with expiring elements lifecycle, Build Log Upload, and Classic Upload. 

I'm wondering if that has anything to do with it. I'm not quite familiar with JobDSL syntax, but that's something to consider. We don't currently support JobDSL, but that's something we should consider.

**Comment by rachely3n** [comment]
Alright, so I've been looking at the GoogleCloudStorageUploader [class](https://github.com/jenkinsci/google-storage-plugin/blob/631adc9c0724bd404012277cfe9e38087d8eb40f/src/main/java/com/google/jenkins/plugins/storage/GoogleCloudStorageUploader.java).

Regarding its usage, I looked at the following (https://github.com/jenkinsci/google-storage-plugin/blob/631adc9c0724bd404012277cfe9e38087d8eb40f/src/test/java/com/google/jenkins/plugins/storage/GoogleCloudStorageUploaderTest.java).

If you see in the test, on this [line](https://github.com/jenkinsci/google-storage-plugin/blob/631adc9c0724bd404012277cfe9e38087d8eb40f/src/test/java/com/google/jenkins/plugins/storage/GoogleCloudStorageUploaderTest.java#L174), a ClassicUpload step is basically running some uploads in a list. I'm not quite sure this can even be expressed in pipeline syntax yet, nevertheless JobDSL.

**Comment by rmeffan** [comment]
Thanks for looking into this further. I should have thought about looking at the test classes...duh. 🙄

Re: Uploads running as a list of upload types, that matches what I saw when I looked at the GoogleCloudStorageUploader class file, and when I checked the config.xml generated by the UI when creating a Post Build Action of type "Build Log Upload" (which maps to class StdoutUpload). My initial description for this issue shows some examples of what I tried, but I was never able to figure out how to represent a list in JobDSL. Didn't try Pipeline syntax, but I thought I might have seen an example that supposedly worked...will have to look at bookmarks and history to see if I can find it.

It sounds like this is something that will need to be implemented vs. just needing documentation to explain how to do it. My Java is rusty, but I could possibly take a stab at it later in June after I have finished some other personal projects. Seems like it will require a mix of learning how the JobDSL plugin works, and what it needs other plugins to do to be supported automagically. Would then need to implement those changes in the GCS plugin. That sound right?

Thanks again for looking at this issue.

**Comment by rachely3n** [comment]
@rmeffan Yes, I'm currently trying to figure out how to just express the list format in pipeline DSL first. It's possible the design of the post build steps will have to change, but it's still in the works. Please hang tight while we try to figure this out!

**Comment by rachely3n** [comment]
Also, an update: https://github.com/jenkinsci/workflow-step-api-plugin/blob/master/README.md#creating-a-basic-synchronous-step

Looks like post build steps can't be run in pipelines as a normal build step. I'm not sure what the case is for JobDSL, but I imagine it's similar. Currently working on making the post build actions steps that can be run as steps which implies you can run them in post.

**Comment by rachely3n** [comment]
If you wanted to stay updated, this is the [PR](https://github.com/jenkinsci/google-storage-plugin/pull/70) I'm currently working on to free some of the post-build steps into pipeline DSL by making them build steps. Should work for JobDSL if it can run normal build steps.

**Comment by rachely3n** [comment]
I've just released 1.3.0. Try it out and let me know how it goes.
Also, provided documentation[ here](https://github.com/jenkinsci/google-storage-plugin/blob/develop/docs/home.md) on using pipeline DSL, I imagine the JobDSL syntax is similar.

So instead of using GoogleCloudStorageUploader, I've made the post-build steps their own steps to use.

**Comment by stephenashank** [comment]
Resolved by #70
